Testing: "class loader confusion 1" (regression 3.25.0->3.25.1 + 3.25.2)

### Describe the bug

We have this version-bump PR https://github.com/apache/polaris/pull/2279 which logs a lot of new test failures for the bump of Quarkus from 3.25.0 to 3.25.2.

The test class `org.apache.polaris.service.task.BatchFileCleanupTaskHandlerTest` yields exceptions like the following for 3.25.1 and 3.25.2:
```

org/apache/polaris/service/catalog/io/FileIOFactory
java.lang.NoClassDefFoundError: org/apache/polaris/service/catalog/io/FileIOFactory
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:566)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:577)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
	at org.apache.polaris.service.task.BatchFileCleanupTaskHandlerTest.buildTaskFileIOSupplier(BatchFileCleanupTaskHandlerTest.java:62)
	at org.apache.polaris.service.task.BatchFileCleanupTaskHandlerTest.testCleanupWithRetries(BatchFileCleanupTaskHandlerTest.java:244)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:1000)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:848)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ClassNotFoundException: org.apache.polaris.service.catalog.io.FileIOFactory
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:550)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
	... 13 more


org.apache.polaris.service.catalog.io.FileIOFactory
java.lang.ClassNotFoundException: org.apache.polaris.service.catalog.io.FileIOFactory
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:550)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:566)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:577)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
	at org.apache.polaris.service.task.BatchFileCleanupTaskHandlerTest.buildTaskFileIOSupplier(BatchFileCleanupTaskHandlerTest.java:62)
	at org.apache.polaris.service.task.BatchFileCleanupTaskHandlerTest.testCleanupWithRetries(BatchFileCleanupTaskHandlerTest.java:244)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:1000)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:848)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

The 2nd exception happens due to the `isBanned` call yielding `true` in `io.quarkus.bootstrap.classloading.QuarkusClassLoader`
```
                if (classPathResourceIndex.isBanned(resourceName)) {
                    throw new ClassNotFoundException(name);
                }
```


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

1. Clone `https://github.com/renovate-bot/apache-_-polaris/tree/renovate/quarkus-platform-and-group`
2. Run/debug `BatchFileCleanupTaskHandlerTest` from e.g. IntelliJ


### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21.0.7

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/actions/BeforeTestAction.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/actions/BeforeTestAction.java
@@ -60,15 +60,14 @@ public class BeforeTestAction implements Action<Task> {
             Test task = (Test) t;
             final Map<String, Object> props = task.getSystemProperties();
 
-            ApplicationModel applicationModel = ToolingUtils
-                    .deserializeAppModel(applicationModelPath.get().getAsFile().toPath());
+            final Path serializedModel = applicationModelPath.get().getAsFile().toPath();
+            ApplicationModel applicationModel = ToolingUtils.deserializeAppModel(serializedModel);
 
             SmallRyeConfig config = effectiveProvider().buildEffectiveConfiguration(applicationModel, new HashMap<>())
                     .getConfig();
             config.getOptionalValue(TEST.getProfileKey(), String.class)
                     .ifPresent(value -> props.put(TEST.getProfileKey(), value));
 
-            final Path serializedModel = applicationModelPath.get().getAsFile().toPath();
             props.put(BootstrapConstants.SERIALIZED_TEST_APP_MODEL, serializedModel.toString());
 
             StringJoiner outputSourcesDir = new StringJoiner(",");

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusApplicationModelTask.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusApplicationModelTask.java
@@ -169,6 +169,7 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
             collectDestinationDirs(module.getMainSources().getSourceDirs(), paths);
             collectDestinationDirs(module.getMainSources().getResourceDirs(), paths);
             appArtifact.setResolvedPaths(paths.build());
+            appArtifact.setReloadable().setWorkspaceModule();
         } else {
             appArtifact.setResolvedPaths(PathList.empty());
         }
@@ -287,7 +288,8 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
                     artifact.type);
             if (!isDependency(artifact)
                     || modelBuilder.getDependency(artifactKey) != null
-                    || modelBuilder.getApplicationArtifact().getKey().equals(artifactKey)) {
+                    // test fixtures depend on the default jar artifact, which could be the root one
+                    || isApplicationRoot(modelBuilder, artifactKey)) {
                 continue;
             }
 
@@ -341,6 +343,10 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
         }
     }
 
+    private static boolean isApplicationRoot(ApplicationModelBuilder modelBuilder, ArtifactKey artifactKey) {
+        return modelBuilder.getApplicationArtifact().getKey().equals(artifactKey);
+    }
+
     private static ModuleVersionIdentifier getModuleVersion(ResolvedDependencyResult resolvedDependency) {
         return Objects.requireNonNull(resolvedDependency.getSelected().getModuleVersion());
     }
@@ -353,10 +359,10 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
 
     private static void collectExtensionDependencies(QuarkusResolvedClasspath classpath, ApplicationModelBuilder modelBuilder) {
         Map<ComponentIdentifier, List<QuarkusResolvedArtifact>> artifacts = classpath.resolvedArtifactsByComponentIdentifier();
-        Set<ArtifactKey> alreadyVisited = new HashSet<>();
+        final Set<ModuleVersionIdentifier> processedModules = new HashSet<>();
         classpath.getRoot().get().getDependencies().forEach(d -> {
             if (d instanceof ResolvedDependencyResult result) {
-                collectExtensionDependencies(result, modelBuilder, artifacts, alreadyVisited, false);
+                collectExtensionDependencies(result, modelBuilder, artifacts, processedModules, false);
             }
         });
     }
@@ -365,8 +371,12 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
             ResolvedDependencyResult resolvedDependency,
             ApplicationModelBuilder modelBuilder,
             Map<ComponentIdentifier, List<QuarkusResolvedArtifact>> resolvedArtifacts,
-            Set<ArtifactKey> alreadyVisited,
+            Set<ModuleVersionIdentifier> processedModules,
             boolean clearReloadableFlag) {
+        final ModuleVersionIdentifier moduleId = getModuleVersion(resolvedDependency);
+        if (!processedModules.add(moduleId)) {
+            return;
+        }
         List<QuarkusResolvedArtifact> artifacts = getResolvedModuleArtifacts(resolvedArtifacts,
                 resolvedDependency.getSelected().getId());
         if (artifacts.isEmpty()) {
@@ -374,14 +384,17 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
         }
 
         final ModuleVersionIdentifier moduleVersionIdentifier = getModuleVersion(resolvedDependency);
+        boolean clearReloadableFlagChildren = clearReloadableFlag;
         for (QuarkusResolvedArtifact artifact : artifacts) {
 
             String classifier = resolveClassifier(moduleVersionIdentifier, artifact.file);
             ArtifactKey artifactKey = ArtifactKey.of(moduleVersionIdentifier.getGroup(), moduleVersionIdentifier.getName(),
                     classifier,
                     artifact.type);
-            if (!alreadyVisited.add(artifactKey)) {
-                return;
+            if (!isDependency(artifact)
+                    // test fixtures depend on the default jar artifact, which could be the root one
+                    || isApplicationRoot(modelBuilder, artifactKey)) {
+                continue;
             }
 
             ResolvedDependencyBuilder dep = modelBuilder.getDependency(artifactKey);
@@ -393,13 +406,15 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
             dep.setDeploymentCp();
             if (clearReloadableFlag) {
                 dep.clearFlag(DependencyFlags.RELOADABLE);
+            } else if (!dep.isReloadable()) {
+                clearReloadableFlagChildren = true;
             }
-            clearReloadableFlag |= !dep.isReloadable();
         }
 
         for (DependencyResult d : resolvedDependency.getSelected().getDependencies()) {
             if (d instanceof ResolvedDependencyResult result) {
-                collectExtensionDependencies(result, modelBuilder, resolvedArtifacts, alreadyVisited, clearReloadableFlag);
+                collectExtensionDependencies(result, modelBuilder, resolvedArtifacts, processedModules,
+                        clearReloadableFlagChildren);
             }
         }
     }

